Hazelcast.Net/Hazelcast.Client/Statistics.cs

```diff
@@ -0,0 +1,296 @@
+                AddNearCacheStats(stats);
+            }
+            catch (Exception e)
+            {
+                Logger.Finest("Can not collect client statistics.", e);
+            }
+            SendStatsToOwner(stats);
+        }
+
+        private void SendStatsToOwner(StringBuilder stats)
+        {
+            var request = ClientStatisticsCodec.EncodeRequest(stats.ToString());
+            try
+            {
+                _client.GetInvocationService().InvokeOnTarget(request, _ownerAddress);
+            }
+            catch (Exception e)
+            {
+                // suppress exception, do not print too many messages
+                if (Logger.IsFinestEnabled())
+                {
+                    Logger.Finest("Could not send stats ", e);
+                }
+            }
+        }
+
+        private void RegisterMetrics()
+        {
+            _gauges.Add("os.committedVirtualMemorySize",
+                () => Process.GetCurrentProcess().VirtualMemorySize64.ToString());
+            _gauges.Add("os.freePhysicalMemorySize", () => EmptyStatValue);
+            _gauges.Add("os.freeSwapSpaceSize", () => EmptyStatValue);
+            _gauges.Add("os.maxFileDescriptorCount", () => EmptyStatValue);
+            _gauges.Add("os.openFileDescriptorCount", () => Process.GetCurrentProcess().HandleCount.ToString());
+            _gauges.Add("os.processCpuTime", () =>
+            {
+                var totalMilliSec = Process.GetCurrentProcess().TotalProcessorTime.TotalMilliseconds;
+                return ((long) totalMilliSec * 1000000).ToString();
+            });
+            _gauges.Add("os.systemLoadAverage", () => EmptyStatValue); //double value
+            _gauges.Add("os.totalPhysicalMemorySize", () => EmptyStatValue);
+            _gauges.Add("os.totalSwapSpaceSize", () => EmptyStatValue);
+
+            _gauges.Add("runtime.availableProcessors", () => Environment.ProcessorCount.ToString());
+            _gauges.Add("runtime.freeMemory", () => EmptyStatValue);
+            _gauges.Add("runtime.maxMemory", () => Process.GetCurrentProcess().MaxWorkingSet.ToInt64().ToString());
+            _gauges.Add("runtime.totalMemory", () => Process.GetCurrentProcess().WorkingSet64.ToString());
+            _gauges.Add("runtime.uptime",
+                () => ((long) (DateTime.Now - Process.GetCurrentProcess().StartTime).TotalMilliseconds).ToString());
+            _gauges.Add("runtime.usedMemory", () => Process.GetCurrentProcess().WorkingSet64.ToString());
+            _gauges.Add("executionService.userExecutorQueueSize", () => EmptyStatValue);
+        }
+
+        private void FillMetrics(StringBuilder stats, ClientConnection ownerConnection)
+        {
+            AddStat(stats, "lastStatisticsCollectionTime", Clock.CurrentTimeMillis());
+            AddStat(stats, "enterprise", "false");
+            AddStat(stats, "clientType", "CSHARP");
+            AddStat(stats, "clientVersion", DllVersion);
```

Comment:
Java client is not sending this. It may be obtained from server side. Actually the client type can also be obtained from server side endpoint.

Reply:
Client will sent its version . Server side shouldn't add data to it.

Reply:
Isn't this redundant though? Clients send their type and version in AuthenticationTask.

Reply:
As rule of thumb; client sends its all stat data and it's not gathered from multiple sources.

Reply:
should we have a PR for Java for this additional field, so that Java is always our reference implementation?